Mathlib/MeasureTheory/SetSemiring.lean

```diff
@@ -275,6 +288,156 @@ lemma sUnion_union_diffFinset₀_of_subset (hC : IsSetSemiring C) (hs : s ∈ C)
+      · simpa [ht2]
+      · simp only [mem_coe, mem_iUnion, exists_prop, ne_eq, id_eq, forall_exists_index, and_imp]
+        intros i hi j x hx h3 h4
+        -- We show i ⊆ s \ ⋃₀ J
+        have ki : i ⊆ s \ ⋃₀ J := by
+          apply hC.sUnion_diffFinset₀_subsets h1.1 h1.2
+          rw [hK1s] at hi
+          exact hi
+        -- We show j ⊆ ⋃₀ K x ⊆ x ∈ J
+        have hx2 : j ⊆ x := by
+          rw [ht1' x hx] at h3
+          exact subset_trans (subset_sUnion_of_mem h3) (hK3 x hx)
+        have kj : j ⊆ ⋃₀ J := hx2.trans <| subset_sUnion_of_mem hx
+        apply disjoint_of_subset ki kj
+        exact disjoint_sdiff_left
+    · intros a ha
+      simp_rw [hK1_of_ne _ (ne_of_mem_of_not_mem ha hJ)]
+      change ∀ t' ∈ (K a).toSet, t' ⊆ a
+      rw [← sUnion_subset_iff]
+      exact hK3 a ha
+    · refine ⟨hC.empty_not_mem_diffFinset₀ h1.1 h1.2, ?_⟩
+      intros a ha
+      rw [ht1' a ha]
+      exact hK4 a ha
+    · simp only [iUnion_iUnion_eq_or_left, ht2, sUnion_union, ht2, K1]
+      simp_rw [apply_ite, hK5, ← hC.diff_sUnion_eq_sUnion_diffFinset₀ h1.1 h1.2, hK5]
+      simp only [↓reduceIte, diff_union_self]
+
+/-- For some `J ⊆ C`, where `C` is a `SetSemiring`, write `⋃₀ J` as the disjoint union
+over `allDiffFinset₀`.-/
+noncomputable def allDiffFinset₀ (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) :=
+  (hC.allDiffFinset₀_props hJ).choose
+
+lemma allDiffFinset₀_disjoint (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) :
+    J.toSet.PairwiseDisjoint (hC.allDiffFinset₀ hJ) :=
+  (Exists.choose_spec (hC.allDiffFinset₀_props hJ)).1
+
+lemma allDiffFinset₀_subsets_semiring (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) (hj : j ∈ J) :
+    (allDiffFinset₀ hC hJ j).toSet ⊆ C :=
+  (Exists.choose_spec (hC.allDiffFinset₀_props hJ)).2.1 _ hj
+
+lemma allDiffFinset₀_subset_semiring (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) :
+    (disjiUnion J (hC.allDiffFinset₀ hJ) (hC.allDiffFinset₀_disjoint hJ)).toSet ⊆ C := by
+  simp only [disjiUnion_eq_biUnion, coe_biUnion, mem_coe, iUnion_subset_iff]
+  exact fun _ ↦ allDiffFinset₀_subsets_semiring hC hJ
+
+lemma  allDiffFinset₀_pairwiseDisjoint' (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) :
+    (⋃ x ∈ J, (hC.allDiffFinset₀ hJ x).toSet).PairwiseDisjoint id :=
+  (Exists.choose_spec (hC.allDiffFinset₀_props hJ)).2.2.1
+
+lemma allDiffFinset₀_pairwiseDisjoint (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) :
+    PairwiseDisjoint (disjiUnion J (hC.allDiffFinset₀ hJ)
+      (hC.allDiffFinset₀_disjoint hJ)).toSet id := by
+  simp only [disjiUnion_eq_biUnion, coe_biUnion, mem_coe]
+  exact allDiffFinset₀_pairwiseDisjoint' hC hJ
+
+lemma allDiffFinset₀_pairwiseDisjoints (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) (hj : j ∈ J) :
+    PairwiseDisjoint (hC.allDiffFinset₀ hJ j).toSet id := by
+  apply PairwiseDisjoint.subset (hC.allDiffFinset₀_pairwiseDisjoint hJ)
+  simp only [disjiUnion_eq_biUnion, coe_biUnion, mem_coe]
+  apply subset_iUnion₂_of_subset j hj (by rfl)
+
+lemma allDiffFinset₀_subset (hC : IsSetSemiring C) (hJ : ↑J ⊆ C) (hj : j ∈ J) :
```

Comment:
Ha, this clashes with the name I suggested above. Maybe `sUnion_allDiffFinset₀_subset_self` for this one.